In [2]:
%load_ext autoreload
%autoreload 2
from datasets import DatasetDict
from pysentimiento.lince.pos import load_datasets, id2label


train_dataset, dev_dataset, test_dataset = load_datasets("es")

lince_pos = DatasetDict(
    train=train_dataset,
    validation=dev_dataset,
    test=test_dataset,
)

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


Reusing dataset lince (/home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589)


  0%|          | 0/3 [00:00<?, ?it/s]

Loading cached processed dataset at /home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589/cache-f8462a766744bd0d.arrow
Loading cached processed dataset at /home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589/cache-b2415cc41a42ddca.arrow
Loading cached processed dataset at /home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589/cache-302bcedfefea246f.arrow
Loading cached processed dataset at /home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589/cache-986aab105c31f86f.arrow
Loading cached processed dataset at /home/jmperez/.cache/huggingface/datasets/lince/pos_spaeng/1.0.0/10d41747f55f0849fa84ac579ea1acfa7df49aa2015b60426bc459c111b3d589/cache-c56998e93fa6679b.arrow
Loading cached processed 

In [3]:
id2label

['VERB',
 'PUNCT',
 'PRON',
 'NOUN',
 'DET',
 'ADV',
 'ADP',
 'INTJ',
 'CONJ',
 'ADJ',
 'AUX',
 'SCONJ',
 'PART',
 'PROPN',
 'NUM',
 'UNK',
 'X']

In [4]:
for ex in lince_pos["train"]:
    assert all(x is not None for x in ex["labels"])

¿Puede pasar que haya etiquetas que no estén en dev o test?

In [5]:
train_tags = set(tag for ex in lince_pos["train"] for tag in ex["pos"])
dev_tags = set(tag for ex in lince_pos["validation"] for tag in ex["pos"])

train_tags == dev_tags

True

In [8]:
lince_pos

DatasetDict({
    train: Dataset({
        features: ['idx', 'words', 'lid', 'pos', 'labels'],
        num_rows: 27893
    })
    validation: Dataset({
        features: ['idx', 'words', 'lid', 'pos', 'labels'],
        num_rows: 4298
    })
    test: Dataset({
        features: ['idx', 'words', 'lid', 'pos', 'labels'],
        num_rows: 10720
    })
})

In [18]:
from pysentimiento.training import load_model
from transformers import AutoModelForTokenClassification
from pysentimiento.lince.pos import label2id, id2label, tokenize_and_align_labels

base_model = "pysentimiento/robertuito-base-uncased"

max_length = 128
model, tokenizer = load_model(
    base_model, label2id=label2id, id2label=id2label,
    max_length=max_length,
    auto_class=AutoModelForTokenClassification
)

tokenized_datasets = lince_pos.map(
    lambda x: tokenize_and_align_labels(x, tokenizer),
    batched=True,
    #remove_columns=lince_ner["train"].column_names,
)



Loading model pysentimiento/robertuito-base-uncased


Some weights of the model checkpoint at pysentimiento/robertuito-base-uncased were not used when initializing RobertaForTokenClassification: ['lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForTokenClassification were not initialized from the model checkpoint at pysentimiento/robertuito-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a dow

  0%|          | 0/28 [00:00<?, ?ba/s]

  0%|          | 0/5 [00:00<?, ?ba/s]

  0%|          | 0/11 [00:00<?, ?ba/s]

Sanity check

In [19]:
from tqdm.auto import tqdm

problematic = []

for split in tokenized_datasets:
    for idx, ex in tqdm(enumerate(tokenized_datasets[split]), total=len(tokenized_datasets[split])):
        if not (len(ex["words"]) == (ex["word_ids"][-2] + 1)):
            problematic.append((split, idx))

print(f"{len(problematic)} problematic examples")


  0%|          | 0/27893 [00:00<?, ?it/s]

  0%|          | 0/4298 [00:00<?, ?it/s]

  0%|          | 0/10720 [00:00<?, ?it/s]

1 problematic examples


In [22]:
len(lince_pos["train"][10310]["words"])

138

Ok, es un problema del largo del tweet (qué ganas de romper las bolas que tienen, 138 palabras!)

In [23]:
assert model.config.num_labels == len(id2label)

In [24]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(tokenizer=tokenizer)



In [29]:
from transformers import TrainingArguments, Trainer
from pysentimiento.lince.pos import compute_metrics

batch_size = 32
epochs = 5

training_args = TrainingArguments(
    output_dir="/tmp/pysentimiento-prueba",
    num_train_epochs=epochs,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    warmup_ratio=0.1,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    do_eval=False,
    weight_decay=0.01,
    logging_dir='./logs',
)

trainer_args = {
    "model": model,
    "args": training_args,
    "compute_metrics": compute_metrics,
    "train_dataset": tokenized_datasets["train"],
    "eval_dataset": tokenized_datasets["validation"],
    "data_collator": data_collator,
    "tokenizer": tokenizer,
}

trainer = Trainer(**trainer_args)

trainer.train()


PyTorch: setting up devices
The default value for the training argument `--report_to` will change in v5 (from all installed integrations to none). In v5, you will need to use `--report_to all` to get the same behavior as now. You should start updating your code and make this info disappear :-).
The following columns in the training set  don't have a corresponding argument in `RobertaForTokenClassification.forward` and have been ignored: word_ids, pos, words, lid, idx.
***** Running training *****
  Num examples = 27893
  Num Epochs = 5
  Instantaneous batch size per device = 32
  Total train batch size (w. parallel, distributed & accumulation) = 32
  Gradient Accumulation steps = 1
  Total optimization steps = 4360


The following columns in the evaluation set  don't have a corresponding argument in `RobertaForTokenClassification.forward` and have been ignored: word_ids, pos, words, lid, idx.
***** Running Evaluation *****
  Num examples = 4298
  Batch size = 32
/home/jmperez/.cache/pypoetry/virtualenvs/pysentimiento-bwlKzHxB-py3.7/lib/python3.7/site-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: INTJ seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/home/jmperez/.cache/pypoetry/virtualenvs/pysentimiento-bwlKzHxB-py3.7/lib/python3.7/site-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: CONJ seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/home/jmperez/.cache/pypoetry/virtualenvs/pysentimiento-bwlKzHxB-py3.7/lib/python3.7/site-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: PRON seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/home/jmperez/.cache/

TrainOutput(global_step=4360, training_loss=0.1496922724837557, metrics={'train_runtime': 613.114, 'train_samples_per_second': 227.47, 'train_steps_per_second': 7.111, 'total_flos': 2300710943790330.0, 'train_loss': 0.1496922724837557, 'epoch': 5.0})